File: src/Entity/Comment.php
<?php

namespace App\Entity;

use App\Repository\CommentRepository;
use Doctrine\Common\Collections\ArrayCollection;
use Doctrine\Common\Collections\Collection;
use Doctrine\DBAL\Types\Types;
use Doctrine\ORM\Mapping as ORM;
use Symfony\Component\Validator\Constraints as Assert;

#[ORM\Entity(repositoryClass: CommentRepository::class)]
class Comment
{
    #[ORM\Id]
    #[ORM\GeneratedValue]
    #[ORM\Column]
    private ?int $id = null;

    #[ORM\Column(type: Types::TEXT)]
    #[Assert\NotBlank(message: 'Comment cannot be empty')]
    #[Assert\Length(min: 2, max: 5000, minMessage: 'Comment must be at least 2 characters', maxMessage: 'Comment cannot exceed 5000 characters')]
    private ?string $content = null;

    #[ORM\Column(length: 180)]
    #[Assert\NotBlank(message: 'Email cannot be empty')]
    #[Assert\Email(message: 'Invalid email format')]
    private ?string $authorEmail = null;

    #[ORM\Column(length: 255)]
    #[Assert\NotBlank(message: 'Author name cannot be empty')]
    #[Assert\Length(min: 2, max: 255, minMessage: 'Author name must be at least 2 characters', maxMessage: 'Author name must not exceed 255 characters')]
    #[Assert\Regex(pattern: "/^[a-zA-ZÀ-ÿ\s\-']+$/", message: 'Author name can only contain letters, spaces, hyphens and apostrophes')]
    private ?string $authorName = null;

    #[ORM\Column(type: Types::DATETIME_MUTABLE)]
    private ?\DateTimeInterface $createdAt = null;

    #[ORM\ManyToOne(inversedBy: 'comments')]
    #[ORM\JoinColumn(nullable: false)]
    private ?BlogPost $blogPost = null;

    #[ORM\ManyToOne(targetEntity: self::class, inversedBy: 'replies')]
    #[ORM\JoinColumn(nullable: true, onDelete: 'CASCADE')]
    private ?self $parentComment = null;

    #[ORM\OneToMany(mappedBy: 'parentComment', targetEntity: self::class, orphanRemoval: true)]
    private Collection $replies;

    public function __construct()
    {
        $this->createdAt = new \DateTime();
        $this->replies = new ArrayCollection();
    }

    public function getId(): ?int
    {
        return $this->id;
    }

    public function getContent(): ?string
    {
        return $this->content;
    }

    public function setContent(string $content): static
    {
        $this->content = $content;
        return $this;
    }

    public function getAuthorEmail(): ?string
    {
        return $this->authorEmail;
    }

    public function setAuthorEmail(string $authorEmail): static
    {
        $this->authorEmail = $authorEmail;
        return $this;
    }

    public function getAuthorName(): ?string
    {
        return $this->authorName;
    }

    public function setAuthorName(string $authorName): static
    {
        $this->authorName = $authorName;
        return $this;
    }

    public function getCreatedAt(): ?\DateTimeInterface
    {
        return $this->createdAt;
    }

    public function setCreatedAt(\DateTimeInterface $createdAt): static
    {
        $this->createdAt = $createdAt;
        return $this;
    }

    public function getBlogPost(): ?BlogPost
    {
        return $this->blogPost;
    }

    public function setBlogPost(?BlogPost $blogPost): static
    {
        if ($this->blogPost === $blogPost) {
            return $this;
        }

        // remove this comment from old blog post's comments collection
        if ($this->blogPost !== null) {
            $this->blogPost->removeComment($this);
        }

        $this->blogPost = $blogPost;

        // add this comment to new blog post's comments collection
        if ($blogPost !== null) {
            $blogPost->addComment($this);
        }

        return $this;
    }

    public function getParentComment(): ?self
    {
        return $this->parentComment;
    }

    public function setParentComment(?self $parentComment): static
    {
        if ($this->parentComment === $parentComment) {
            return $this;
        }

        // remove this from old parent's replies
        if ($this->parentComment !== null) {
            $this->parentComment->removeReply($this);
        }

        $this->parentComment = $parentComment;

        // add this to new parent's replies
        if ($parentComment !== null) {
            $parentComment->addReply($this);
        }

        return $this;
    }

    /**
     * @return Collection<int, self>
     */
    public function getReplies(): Collection
    {
        return $this->replies;
    }

    public function addReply(self $reply): static
    {
        if (!$this->replies->contains($reply)) {
            $this->replies->add($reply);
            $reply->setParentComment($this);
        }
        return $this;
    }

    public function removeReply(self $reply): static
    {
        if ($this->replies->removeElement($reply)) {
            if ($reply->getParentComment() === $this) {
                $reply->setParentComment(null);
            }
        }
        return $this;
    }
}

File: src/main/java/tn/esprit/entities/Comment.java
package tn.esprit.entities;

import java.sql.Timestamp;

public class Comment {
    private int id;
    private String content;
    private String authorEmail;
    private String authorName;
    private Timestamp createdAt;
    private int blogPostId;
    private Integer parentCommentId;

    public Comment() {}

    public Comment(int id, String content, String authorEmail, String authorName,
                   Timestamp createdAt, int blogPostId, Integer parentCommentId) {
        this.id = id;
        this.content = content;
        this.authorEmail = authorEmail;
        this.authorName = authorName;
        this.createdAt = createdAt;
        this.blogPostId = blogPostId;
        this.parentCommentId = parentCommentId;
    }

    public Comment(String content, String authorEmail, String authorName, int blogPostId) {
        this.content = content;
        this.authorEmail = authorEmail;
        this.authorName = authorName;
        this.blogPostId = blogPostId;
    }

    public int getId() { return id; }
    public void setId(int id) { this.id = id; }

    public String getContent() { return content; }
    public void setContent(String content) { this.content = content; }

    public String getAuthorEmail() { return authorEmail; }
    public void setAuthorEmail(String authorEmail) { this.authorEmail = authorEmail; }

    public String getAuthorName() { return authorName; }
    public void setAuthorName(String authorName) { this.authorName = authorName; }

    public Timestamp getCreatedAt() { return createdAt; }
    public void setCreatedAt(Timestamp createdAt) { this.createdAt = createdAt; }

    public int getBlogPostId() { return blogPostId; }
    public void setBlogPostId(int blogPostId) { this.blogPostId = blogPostId; }

    public Integer getParentCommentId() { return parentCommentId; }
    public void setParentCommentId(Integer parentCommentId) { this.parentCommentId = parentCommentId; }

    @Override
    public String toString() {
        return "Comment{id=" + id + ", author='" + authorName + "', postId=" + blogPostId + "}";
    }
}
